### DISCUSSION TITLE
Issue using "checkInside" function

### DISCUSSION THREAD

**Discussion by angus0997** [mtex-toolbox/mtex#1382] [Ask Anything]
Hi all,
I am having an issue using the checkInside function.
I have identified a subset of grains within my data (which are small and circular), and want to check if they are inside the other grains. This works perfectly, except if the small circular grain is on a grain boundary which happens to be a twin boundary. In this case, the small circular grain is counted as being inside one of the grains sharing the twin boundary, even though it is actually at this boundary.

Here are some images to clarify the above:
Here is the full dataset shown as an IPF map. Twin boundaries are drawn in red.
![image](https://user-images.githubusercontent.com/101588689/158226190-f29f1ff1-3873-4490-a2ec-81f798bdc68f.png)

Here, the small circular grains have been identified, and coloured red.
![image](https://user-images.githubusercontent.com/101588689/158226388-7c93a213-5c53-460c-8641-fdec97793a0b.png)

Finally, here, the full dataset with small circular grains erased (coloured white) if they are inside another grain
![image](https://user-images.githubusercontent.com/101588689/158226574-0109eaa8-c6dd-4609-9f79-f23b0494827a.png)

You can notice at the twin boundary towards the top of the final image, one of the small circular grains has been erased, even though it lies on the twin boundary - and so should not be erased! Other examples of this can be found elsewhere in the maps.

Has anyone else experienced this issue, or is it possible I am making a mistake somewhere?
Cheers,
Angus

**Comment by kilir** [comment]
Hi,
could you also supply the relevant lines of script you use to accomplish this?
Cheers,
Rüdiger

**Comment by angus0997** [comment]
Hi,
Sure, here are (I think) all the relevant lines:

%% define grains and twins

% reconstruct the grain structure
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'angle',10*degree);
% remove some very small grains
ebsd(grains(grains.grainSize<5)) = [];
% redo grain segementation
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',10*degree);
% smooth grain boundaries
grains = smooth(grains,5);

%Define twins
cs= ebsd('1').CS;
gB = grains.boundary('1','1');
twin = angle(gB.misorientation,CSL(3,cs))<8.66*degree;
twinBoundary = gB(twin);

% Define colour key for plotting IPF map
ipfKey = ipfColorKey(ebsd('1'));
ipfKey.inversePoleFigureDirection = vector3d.Y;
colors = ipfKey.orientation2color(ebsd('1').orientations);

%% To identify the small circular grains (which are grains of gamma prime phase)
cond_area = grains.area < 17;
Feret_diameters = norm(caliper(grains,'longest'));
cond_circularity = 0.8*(grains.area) < pi/4*(Feret_diameters.^2) & pi/4*(Feret_diameters.^2) < 1.6*(grains.area);

gamma_prime = grains(logical(cond_area.*cond_circularity));     % These are the small circular grains
gamma = grains(~logical(cond_area.*cond_circularity));     % These are all other grains

%% Check if gamma prime are inside gamma grains, or at the boundary
inside = checkInside(gamma,gamma_prime);     % checks for gamma prime in gamma grains
internal_gamma_prime = gamma_prime(any(inside,2));     % Picks out the gamma prime which are inside gamma grains

%% Now make a plot to show the gamma prime which are inside the gamma grains, by plotting them white

figure, plot(ebsd('1'),colors);
% and on top the grain boundaries
hold on
plot(grains.boundary,'linewidth',1)
hold on
plot(gB(twin),'linewidth',1.5,'linecolor','r')
hold on
plot(internal_gamma_prime,'FaceColor','w')

This last plot produces the third figure in the first post. Mostly it is correct, but it appears that some gamma prime which are at the grain boundaries are counted as being "internal_gamma_prime", and hence appear white in this figure.

Thanks,
Angus

**Comment by kilir** [comment]
Hi,
is there any difference with `inside = checkInside(grains,gamma_prime); ` or simply checking all grains to be inclusions of others?
Cheers,
Rüdiger

**Reply by angus0997** [reply]
When using:  inside = checkInside(grains,gamma_prime); 
The result is the same, shown here:
![image](https://user-images.githubusercontent.com/101588689/158568703-620a65ab-a623-43f0-9cfc-c1d67364b98c.png)

To check all grains to be inclusions of others: 
` inside = checkInside(grains,grains); `
` internal_gamma_prime = grains(any(inside,2));  `% Now we select all internal grains, not just the small round ones.

Checking all grains to be inclusions of others seems to help in understanding. It produces this result:
![image](https://user-images.githubusercontent.com/101588689/158569529-6d3ee595-2994-4734-9b85-0b8139b2af7f.png)

Here I can see two examples of grains which don't look to be inclusions, being counted as inclusions. One at the top of the map, and one at the bottom of the map. Both of these grains share a border with one of the incorrectly labelled gamma prime from my initial question. So now my question is, why are these two grains counted as inclusions?

**Reply by kilir** [reply]
If the white grains from your last plot are the ones identified as inclusions by `checkInside(grains,grains)` I'd also think that the result is not correct. Could you provide a little crop of e.g. the topmost part of the map with the incorrectly identified grain (grains&ebsd as a .mat should be sufficient).
Cheers,
Rüdiger
